pkg/deployment/k8s/deployment.go

```diff
@@ -0,0 +1,227 @@
+		ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: d.namespace},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Protocol:   corev1.ProtocolTCP,
+					Port:       int32(port),
+					TargetPort: intstr.FromInt(int(targetPort)),
+				},
+			},
+			Type:     corev1.ServiceTypeClusterIP,
+			Selector: map[string]string{"app": dataplaneAppName},
+		},
+	}
+
+	err := d.client.Update(context.Background(), serviceSpec)
+	if err != nil {
+		d.logger.Errorf("error occurred while updating K8s service %s:", err)
+		return err
+	}
+
+	d.logger.Infof("Updating K8s service at %s:%d ", name, port)
+	return nil
+}
+
+// CreateEndpoint creates a K8s endpoint.
+func (d *Deployment) CreateEndpoint(name, targetIP string, targetPort uint16) error {
+	endpointSpec := &corev1.Endpoints{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: d.namespace,
+		},
+		Subsets: []corev1.EndpointSubset{
+			{
+				Addresses: []corev1.EndpointAddress{
+					{
+						IP: targetIP, // Replace with the desired IP address of the endpoint.
+					},
+				},
+				Ports: []corev1.EndpointPort{
+					{
+						Port: int32(targetPort),
+					},
+				},
+			},
+		},
+	}
+
+	err := d.client.Create(context.Background(), endpointSpec)
+	if err != nil {
+		d.logger.Errorf("error occurred while creating K8s endpoint %s:", err)
+		return err
+	}
+
+	d.logger.Infof("Creating K8s endPoint at %s:%d that connected to external IP: %s:%d", name, targetPort, targetIP, targetPort)
+	return nil
+}
+
+// UpdateEndpoint creates a K8s endpoint.
+func (d *Deployment) UpdateEndpoint(name, targetIP string, targetPort uint16) error {
+	endpointSpec := &corev1.Endpoints{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: d.namespace,
+		},
+		Subsets: []corev1.EndpointSubset{
+			{
+				Addresses: []corev1.EndpointAddress{
+					{
+						IP: targetIP, // Replace with the desired IP address of the endpoint.
+					},
+				},
+				Ports: []corev1.EndpointPort{
+					{
+						Port: int32(targetPort),
+					},
+				},
+			},
+		},
+	}
+
+	err := d.client.Update(context.Background(), endpointSpec)
+	if err != nil {
+		d.logger.Errorf("error occurred while updating K8s endpoint %s:", err)
+		return err
+	}
+
+	d.logger.Infof("Updating K8s endPoint at %s:%d that connected to external IP: %s:%d", name, targetPort, targetIP, targetPort)
+	return nil
+}
+
+// DeleteEndpoint deletes a k8s endpoint.
+func (d *Deployment) DeleteEndpoint(name string) error {
+	endpointSpec := &corev1.Endpoints{
+		ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: d.namespace}}
+
+	err := d.client.Delete(context.Background(), endpointSpec)
+	if err != nil {
+		d.logger.Errorf("error occurred while deleting K8s endpoint %s:", err)
+		return err
+	}
+
+	d.logger.Infof("Deleting K8s endPoint %s", name)
+	return nil
+}
+
+// GetPodLabelsByIP returns all the labels that match the pod IP.
+func (d *Deployment) GetPodLabelsByIP(podIP string) (map[string]string, error) {
```

Comment:
In the earlier version of control plane, we had used kube informers.
I am unsure of the latency of this call since we have to get the entire pod list and iterate through it upon every incoming connection, which is very inefficient. Can you try using informers which are synced and kept in memory (https://github.com/clusterlink-net/clusterlink/blob/main/pkg/k8s/kubernetes/kubeinformer.go) . I am unaware of any downsides to this approach.

Reply:
I agree, but the main point here of this change is to use more high-level abstraction I will check if we can cache the pod lists also here. @elevran any suggestions.

Reply:
Informers can be used and would work. I was suggesting to use a higher level interface (via the `controller-runtime` packages) but to make real use of it, it would run a control loop that is notified of Pod change events and then cache the needed object attributes on these changes. See, for example, the following example of using informers [here](https://github.com/elevran/podip) along with the needed code to cache objects (e.g., validate the Pod was assigned an IP, handle IPv4 and IPv6, etc.)

Reply:
I put this code in different PR